controllers/hardwareclassification_controller.go

```diff
@@ -33,24 +41,82 @@ type HardwareClassificationReconciler struct {
 
 // Add RBAC rules to access baremetalhost resources
 // +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts,verbs=get;list;update
 // +kubebuilder:rbac:groups=metal3.io,resources=baremetalhosts/status,verbs=get
 
-func (r *HardwareClassificationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("hardwareclassification", req.NamespacedName)
+func (hcReconciler *HardwareClassificationReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Initialize the logger with namespace
+	hcReconciler.Log = hcReconciler.Log.WithName(HWControllerName).WithValues("metal3-harwdwareclassification", req.NamespacedName)
+
+	// Get HardwareClassificationController to get values for Namespace and ExpectedHardwareConfiguration
+	hardwareClassification := &hwcc.HardwareClassification{}
+
+	if err := hcReconciler.Client.Get(ctx, req.NamespacedName, hardwareClassification); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Get ExpectedHardwareConfiguraton from hardwareClassification
+	extractedProfile := hardwareClassification.Spec.HardwareCharacteristics
+	hcReconciler.Log.Info("Extracted hardware configurations successfully", "Profile", extractedProfile)
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(hardwareClassification, hcReconciler.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the hardwareClassification object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, hardwareClassification); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Get the new hardware classification manager
+	hcManager := utils.NewHardwareClassificationManager(hcReconciler.Client, hcReconciler.Log)
+
+	ErrValidation := hcManager.ValidateExtractedHardwareProfile(extractedProfile)
+	if ErrValidation != nil {
+		hcReconciler.Log.Error(ErrValidation, ErrValidation.Error())
+		hcReconciler.handleErrorConditions(req, hardwareClassification, hwcc.ProfileMisConfigured, ErrValidation.Error(), hwcc.ProfileMatchStatusEmpty)
+		return ctrl.Result{}, nil
+	}
+
+	//Fetch baremetal host list for the given namespace
+	hostList, _, err := hcManager.FetchBmhHostList(hardwareClassification.ObjectMeta.Namespace)
+	if err != nil {
+		errMessage := "Unable to fetch BMH list from BMO"
+		hcReconciler.handleErrorConditions(req, hardwareClassification, hwcc.FetchBMHListFailure, errMessage, hwcc.ProfileMatchStatusEmpty)
+		return ctrl.Result{}, nil
+	}
+
+	if len(hostList) == 0 {
+		errMessage := "No BareMetalHost found in ready state"
+		hcReconciler.Log.Info("No BareMetalHost found in ready state")
+		hcReconciler.handleErrorConditions(req, hardwareClassification, hwcc.NoBMHHost, errMessage, hwcc.ProfileMatchStatusEmpty)
```

Comment:
Do you really want to treat this as an error ? Isn't it an expected state ?

Reply:
@maelk We have changed the code for it, its just the the info instead of error.
Kindly review commit id e124097e7e8313c08b5c95ca95b1952c089a50a4